Preparations for debugging shapes

spiessimon opened the pull request:
This PR prepares for #4398. It adds a new flag `-shape-format (old-merlin|debugging-shapes)` to the compiler, which switches between the default shapes for Merlin and (eventually) an extended version of the shapes that contain more information for debugging. In the current version, the shapes that we produce for debugging are only dummies. In subsequent PRs, we will enrich them, make sure that Merlin still works on them, and then eventually switch over to debugging-shapes as the default. 


**Review.**

There are two key aspects in reviewing this PR:
1. making sure that the new shapes are not yet enabled by default and instead their generation is guarded by the flag mentioned above.
2. making sure that moving down the shapes in the typing of type declarations does not affect Merlin (or other clients of the shapes).

liam923 commented on typing/typedecl.ml:
```diff
@@ -1076,6 +1111,10 @@ let transl_declaration env sdecl (id, uid) =
       in
       set_private_row env sdecl.ptype_loc p decl
     end;
+    (* CR sspies: We used to compute shapes here, which were then added to
+      various typing environments. The computation of the shapes has moved
+      further down in the translation, so they are currently not added to the
+      intermediate environments. Find out whether that is an issue.   *)
```
I did some testing and this doesn't seem to break Merlin - see internal feature fe-1775906

spiessimon replied:
Perfect!

liam923 commented on typing/typedecl.ml:
```diff
@@ -1786,8 +1815,15 @@ let update_constructor_representation
 
-let add_types_to_env decls shapes env =
-  List.fold_right2
+let add_types_to_env ?shapes decls env =
```
I'd rather make `shapes` a required param.

spiessimon replied:
Done.

liam923 reviewed (changes requested):
lgtm, modulo the small things I mentioned.

spiessimon commented on driver/oxcaml_args.ml:
```diff
@@ -797,10 +797,13 @@ let mk_restrict_to_upstream_dwarf f =
     " Only emit the same DWARF information as the upstream compiler" )
 
+(* CR sspies: This helper text must be updated when -shape-format is eventually
+   removed. *)
```
Good point. I removed this one and added one for `-shape-format`.

spiessimon commented:
I added the field `cms_shape_format`, which always stores the value of the command line flag. I also removed the implication between `gno-upstream-dwarf` and `shape-format debugging-shapes` for now to avoid accidental triggering of the new shapes. This can be re-enabled once Merlin support is fixed (see https://github.com/oxcaml/oxcaml/pull/4440).

liam923 commented:
Can you also make [here](https://github.com/spiessimon/oxcaml/blob/debugging-uids-single-shape-type-preparation/tools/objinfo.ml#L277) print out the shape format?

liam923 reviewed (approved):
Thanks
